File: Lab2/C#/MyText.cs
using System;
using System.Linq;

namespace Lab2
{
    public class MyText
    {
        private MyString[] Value;
        
        public MyText()
        {
            this.Value = new MyString[0];
        }

        public void AddString(MyString myString)
        {
            Array.Resize(ref Value, Value.Length + 1);
            Value[^1] = myString;
        }

        public void Write()
        {
            foreach (var myString in this.Value)
            {
                foreach (var character in myString.GetValue())
                {
                    Console.Write(character);
                }

                Console.Write(' ');
            }

            Console.WriteLine();
        }

        public void RemoveString(int index)
                 {
                     var temp = new MyString[Value.Length - 1];
                     for (var i = 0; i < Value.Length; i++)
                     {
                         if (i < index)
                         {
                             temp[i] = Value[i];
                         }
         
                         if (i == index) { continue; }
                         if (i > index)
                         {
                             temp[i-1] = Value[i];
                         }
                     }
                     
                     Value = temp;
                 }

        public void EraseText()
        {
            this.Value = new MyString[0];
        }

        public double AverageLength()
        {
            if (Value.Length == 0) { return 0; }

            var result = (double) this.Value.Sum(myString => myString.GetValue().Length) / this.Value.Length;
            return result;
            
        }

        public double PercentOfVowel()
        {
            double letters = 0;
            double vowel = 0;
            foreach (var array in this.Value)
            {
                foreach (var letter in array.GetValue())
                {
                    letters++;
                    if ("eyuioaEYUIOA".Contains(letter))
                    {
                        vowel++;
                    }
                }
            }

            return (vowel / letters) * 100;
        }

        public void RemoveString(string substring)
        {
            for (var index = 0; index < this.Value.Length; index++)
            {
                var str = this.Value[index];
                for (var i = 0; i < str.GetValue().Length; i++)
                {
                    if (str.GetValue()[i] != substring[0]) continue;
                    var counter = 0;
                    for (var x = 0; x < substring.Length; x++)
                    {
                        if (str.GetValue()[i + x] != substring[x])
                        {
                            break;
                        }

                        counter++;
                    }

                    if (counter != substring.Length) continue;
                    RemoveString(index);
                    index--;
                }
            }
        }
        
    }
}

File: Lab2/C++/MyText.cpp
#include <iostream>
#include <cstring>
#include <vector>
#include "MyText.h"

using namespace std;

void MyText::addString(MyString newString) {
    text.push_back(newString);
    this->size++;
};

void MyText::write() {
   for (MyString i : text) {
       cout << i.getValue() << " ";
   }
   cout << endl;
}

void MyText::removeString(int index) {
    text.erase(text.begin() + index ) ;
};

void MyText::removeString(MyString substring) {
    if (substring.getLength() > 0) {
        char key = substring.getValue()[0];
        for (int i = getLength()-1; i >= 0; i--) {
            for (int j = 0; j < text[i].getLength(); j++) {
                if (text[i].getValue()[j] == key && j + substring.getLength() <= text[i].getLength()) {
                    bool found = true;
                    for (int k = j; k < j+substring.getLength(); k++) {
                        if (text[i].getValue()[k] != substring.getValue()[k - j]) {
                            found = false;
                            break;
                        }
                    }
                    if (found) {
                        removeString(i);
                        break;
                    }
                }
            }
        }
    }
}

double MyText::averageLength() {
    float result;
    int length = 0;
    for (MyString str : text) {
        length += str.getLength();
    }
    result =  length / text.size();
    return result;
}

double MyText::percentOfVowel() {
    const std::string vowel_letters("eyuioa");
    int letters = 0;
    int vowel = 0;
    for (MyString str : text) {
        letters += str.getLength();
        for (int i = 0; i < str.getLength(); i++) {
            if (isVowel(str.getValue()[i])) {
                vowel += 1;
            }
        }
    }
    double percent = (double) vowel / letters * 100;
    return percent;
};

void MyText::eraseText() {
    text.erase(text.begin(),text.end());
}

bool MyText::isVowel(char a) {
    return a == 'e' || a == 'y' || a == 'u' || a == 'i' || a == 'o' || a == 'a' ||
        a == 'E' || a == 'Y' || a == 'U' || a == 'I' || a == 'O' || a == 'A';
}

int MyText::getLength() {
    return this->size;
}
